Replace SourcePawn grammar

Sarrus1 opened the pull request:
<!--- Briefly describe your changes in the field above. -->
Replace SourcePawn grammar

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->
Dreae's grammar is no longer maintained and is missing a few Sourcepawn core features. This PR replaces it with mine, which now properly handles the old and new syntax, most importantly type highlighting for variable declarations.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Please remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] **I am changing the source of a syntax highlighting grammar**
  - Old: https://github.com/Dreae/sourcepawn-vscode
  - New: https://github.com/Sarrus1/sourcepawn-vscode/tree/main/editors/code

Sarrus1 commented:
A couple more things:

## 1

I would also like to point out my fully functional tree-sitter grammar for Sourcepawn [here](https://github.com/Sarrus1/tree-sitter-sourcepawn). 
It has highlighting queries, as well as tags queries for Github navigation support.
From my understanding of the discussion [here](https://github.com/github-linguist/linguist/discussions/6073#discussion-4394815), it is not possible to contribute tree-sitter grammars to linguist. Is this still the case? Is there a way to get the tree-sitter grammar into Github by any chance?

## 2
The textmate grammar supports JSDoc injections, is there something I should do to add support to that in Linguist as well?
https://github.com/Sarrus1/sourcepawn-vscode/blob/main/editors/code/syntaxes/sourcepawn-jsdoc.json

## 3
This repo can be deleted now: https://github.com/github-linguist/sublime-sourcepawn

Thanks for taking the time to consider this PR!

lildude commented:
> Is this still the case? 

Yes

> Is there a way to get the tree-sitter grammar into Github by any chance?

No

> The textmate grammar supports JSDoc injections, is there something I should do to add support to that in Linguist as well?

I'm not sure what there would be to add. Linguist only acts as a collection point which bundles them up for the syntax highlighting engine which is what reads the grammars. No processing of the files takes place other than some basic PCRE regex checking and then conversion to JSON if not already in JSON format. We expect the grammar file to be in its final and complete state as no other processing takes place.

lildude reviewed (changes requested):
Please use the `script/add-grammar` script to replace the grammar as documented in the CONTRIBUTING.md file.

Sarrus1 reviewed (commented):
Done 👍

lildude reviewed (commented):
🤔 there should be more changes than this. Your repo has several grammar scope names so the `grammars.yml` file should have been updated to include these. The cached license file should also have changed as the commits are different. Are you sure you've added all the changed files?

Sarrus1 commented:
> 🤔 there should be more changes than this. Your repo has several grammar scope names so the `grammars.yml` file should have been updated to include these. The cached license file should also have changed as the commits are different. Are you sure you've added all the changed files?

I reran the script just to be sure, here are the commands I ran:

```bash
> script/add-grammar --replace "sourcepawn-vscode" https://github.com/Sarrus1/sourcepawn-vscode
Checking Docker is installed and running
Deregistering submodule: vendor/grammars/sourcepawn-vscode
Cleared directory 'vendor/grammars/sourcepawn-vscode'
Submodule 'vendor/grammars/sourcepawn-vscode' (https://github.com/Dreae/sourcepawn-vscode) unregistered for path 'vendor/grammars/sourcepawn-vscode'
rm 'vendor/grammars/sourcepawn-vscode'
Registering new submodule: https://github.com/Sarrus1/sourcepawn-vscode.git
Reactivating local git directory for submodule 'vendor/grammars/sourcepawn-vscode'
latest: Pulling from linguist/grammar-compiler
Digest: sha256:fcaf08a6e27d3d88212d9ca8aceee30d4089499a77bb05307871febd8073790f
Status: Image is up to date for linguist/grammar-compiler:latest
docker.io/linguist/grammar-compiler:latest

What's Next?
  1. Sign in to your Docker account → docker login
  2. View a summary of image vulnerabilities and recommendations → docker scout quickview linguist/grammar-compiler:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
OK! added grammar source 'vendor/grammars/sourcepawn-vscode'
        new scope: source.sourcepawn
Caching grammar license
Caching dependency records for linguist
  git_submodule
    Using AL (96c407663984341b21e574a58ef6ca6f89cc3fc0)
    Using Alloy.tmbundle (0a80fc74e921325da23264889771a88a02877414)
    Using Assembly-Syntax-Definition (519b7b3c5f8428832dde72a90041d0707f831e5c)
    .... # cropping some of it
  * 451 git_submodule dependencies
Updating grammar documentation in vendor/README.md

> git status
On branch master
Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   .gitmodules

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .gitmodules
	modified:   Brewfile
	modified:   vendor/README.md

> git add --all
> git commit -m "fix: replace SourcePawn grammar"
> git push -f
```

The grammar is located in [this folder](https://github.com/Sarrus1/sourcepawn-vscode/tree/main/editors/code/syntaxes) which has a parent directory with an MIT License. Maybe the location is a problem?

Sarrus1 commented:
Hello @lildude I think I finally managed to make the script work properly. All the changes should be in order now 👍

Sarrus1 commented:
Hello, bumping this. Is there anything else I can do to improve this PR?

lildude reviewed (changes requested):
Sorry I missed this earlier, but we can not accept this change.

Whilst the grammar files are licensed under MIT, the over-all repo is licensed by GPL-3 which means we'll potentially be shipping code that is covered by this license which we can't do.

I'm also not sure GPL3 allows sub-licensing.

Sarrus1 commented:
Should I move my grammar to a new repo to solve the issue?

lildude commented:
That would probably be best short of relicensing the current repo, but that's entirely up to you. We can't accept the current repo as it stands with the GPL-3 license.

Sarrus1 commented:
Changing the License to MIT is fine with me 👍 
I have changed it on the [main branch](https://github.com/Sarrus1/sourcepawn-vscode/tree/main) and [feat/salsa](https://github.com/Sarrus1/sourcepawn-vscode/tree/feat/salsa) which is currently the default branch until its merged. 
Should I rerun something to account for that?

lildude commented:
> Should I rerun something to account for that?

Yes. Re-running `script/add-grammar --replace...` like you did when switching the grammars should pull in the updated license file.

Sarrus1 commented:
All done 👍

lildude commented on grammars.yml:
```diff
@@ -1176,9 +1180,6 @@ vendor/grammars/vscode-singularity:
 - source.slice
-vendor/grammars/vscode-vba:
-- source.vba
-- source.wwb
```
You didn't want to do this 😁

Sarrus1 replied:
I just ran the script, no idea how that happened, reverted 👍

lildude commented:
The changes to `vendor/licenses/git_submodule/sourcepawn-vscode.dep.yml` are also not consistent with what I'd have expected the various scripts to have done. This looks like a manual change.

Sarrus1 commented:
> This looks like a manual change.

I did the manual change in a previous commit and just reran the script thinking it would overwrite it.

I have now reran the script again and made sure that it edited the files correctly.

lildude commented on vendor/README.md:
```diff
@@ -1,5 +1,4 @@
-Grammar index
-=============
+# Grammar index
```
This file is automatically maintained. Do not make changes directly to it.

lildude replied:
NM. I see I'm looking at old commits.

lildude commented on vendor/README.md:
```diff
@@ -572,6 +571,7 @@ This is a list of grammars that Linguist selects to provide syntax highlighting
 - **UnrealScript:** [textmate/java.tmbundle](https://github.com/textmate/java.tmbundle)
 - **UrWeb:** [gwalborn/UrWeb-Language-Definition](https://github.com/gwalborn/UrWeb-Language-Definition)
 - **V:** [0x9ef/vscode-vlang](https://github.com/0x9ef/vscode-vlang)
+- **VBA:** [serkonda7/vscode-vba](https://github.com/serkonda7/vscode-vba)
```
You shouldn't need to make this change yourself... you're only changing the source of the SourcePawn grammar, not anything else.

lildude replied:
NM. I see I'm looking at old commits.

Sarrus1 commented:
I'm not sure what I should do to get the script to recognize that the repo has an MIT license instead of "others".

lildude commented on vendor/licenses/git_submodule/sourcepawn-vscode.dep.yml:
```diff
@@ -1,15 +1,15 @@
-homepage: https://github.com/Dreae/sourcepawn-vscode
-license: mit
+homepage: https://github.com/Sarrus1/sourcepawn-vscode.git
+license: other
```
I've taken a look and I think it's because the licensed gem we use isn't detecting the file as MIT for some reason so we'll need to manually override it.

```suggestion
license: mit
```

Sarrus1 commented:
I think everything should be in order now! The commit version is correct :+1:

lildude reviewed (approved):
LGTM. Thanks.

Note: this PR will not be merged until close to when the next release is made. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.
